pandas/core/generic.py

```diff
@@ -726,7 +726,16 @@ def transpose(self, *args, **kwargs):
             new_values = new_values.copy()
 
         nv.validate_transpose(tuple(), kwargs)
-        return self._constructor(new_values, **new_axes).__finalize__(self)
+        result = self._constructor(new_values, **new_axes).__finalize__(self)
+
+        if self.ndim == 2 and self._is_homogeneous_type and len(self.columns):
```

Comment:
I think its worthwhile to make a method to encapsulate this maybe

def _homogeneous_dtype(self):
   # return the single dtype if homogeneous, None if not